xgcm/test/test_grid_ufunc.py

```diff
@@ -81,6 +83,112 @@ def test_invalid_signatures(self, signature):
+
+    def test_annotated_args(self):
+        # TODO test hints without annotations
+        # TODO test hints with annotations that don't conform to Xgcm
+
+        @as_grid_ufunc()
+        def ufunc(a: Gridded[np.ndarray, "X:center"]):
+            ...
+
+        assert ufunc.signature == "(X:center)->()"
+
+        @as_grid_ufunc()
+        def ufunc(a: Gridded[np.ndarray, "X:center,Y:center"]):
+            ...
+
+        assert ufunc.signature == "(X:center,Y:center)->()"
+
+        @as_grid_ufunc()
+        def ufunc(
+            a: Gridded[np.ndarray, "X:left"],
+            b: Gridded[np.ndarray, "Y:right"],
+        ):
+            ...
+
+        assert ufunc.signature == "(X:left),(Y:right)->()"
+
+    def test_annotated_return_args(self):
+        @as_grid_ufunc()
+        def ufunc() -> Gridded[np.ndarray, "X:center"]:
+            ...
+
+        assert ufunc.signature == "()->(X:center)"
+
+        @as_grid_ufunc()
+        def ufunc() -> Gridded[np.ndarray, "X:left,Y:right"]:
+            ...
+
+        assert ufunc.signature == "()->(X:left,Y:right)"
+
+        @as_grid_ufunc()
+        def ufunc() -> Tuple[
+            Gridded[np.ndarray, "X:left"], Gridded[np.ndarray, "Y:right"]
+        ]:
+            ...
+
+        assert ufunc.signature == "()->(X:left),(Y:right)"
+
+        @as_grid_ufunc()
+        def ufunc(
+            a: Gridded[xr.DataArray, "X:center"]
+        ) -> Gridded[np.ndarray, "X:left"]:
+            ...
+
+        assert ufunc.signature == "(X:center)->(X:left)"
+
+    @pytest.mark.xfail
+    def test_invalid_arg_annotation(self):
+        with pytest.raises(
+            ValueError, match="Argument a has an invalid grid signature"
+        ):
+
+            @as_grid_ufunc()
+            def ufunc(a: Gridded[np.ndarray, "(X:Mars)"]):
+                ...
+
+    @pytest.mark.xfail
+    def test_invalid_return_arg_annotation(self):
+        with pytest.raises(
+            ValueError, match="Return argument(s) have an invalid grid signature"
+        ):
+
+            @as_grid_ufunc()
+            def ufunc() -> Gridded[np.ndarray, "(X:Venus)"]:
+                ...
+
+    def test_both_sig_kwarg_and_hints_given(self):
+        with pytest.raises(
+            ValueError, match="only one of either type hints or signature kwarg"
+        ):
+
+            @as_grid_ufunc(signature="(X:center)->(X:left)")
+            def ufunc(
+                a: Gridded[np.ndarray, "(X:center)"]
+            ) -> Gridded[np.ndarray, "(X:left)"]:
+                ...
+
+    def test_type_hint_as_numpy_ndarray(self):
+        # TODO I want this to fail mypy but it doesn't
+        @as_grid_ufunc()
+        def ufunc1(a: Gridded[str, "(X:center)"]):
+            ...
+
+        # This should pass mypy
+        @as_grid_ufunc()
+        def ufunc3(a: Gridded[np.ndarray, "(X:center)"]):
+            # np.ndarray has a .strides method but str doesn't (and nor does xr.DataArray)
+            print(a.strides)
```

Comment:
What do I need to do to get this to fail mypy? Is it because [`xarray.apply_ufunc` allows its arguments to be `Any`](https://github.com/pydata/xarray/blob/33067cd24f66d4855babaa6801b009480c4e2cb2/xarray/core/computation.py#L804)?

Reply:
I tried narrowing the `Any` upstream, but I think my debugging is being thwarted by pre-commit behaving unexpectedly again:

```
(py39) tom@pop-os:~/Documents/Work/Code/xgcm$ mypy xgcm
xgcm/grid_ufunc.py:399: note: Revealed type is "builtins.tuple[xarray.core.dataarray.DataArray]"
xgcm/grid_ufunc.py:554: note: Revealed type is "builtins.list[xarray.core.dataarray.DataArray]"
(py39) tom@pop-os:~/Documents/Work/Code/xgcm$ pre-commit run --all-files
Trim Trailing Whitespace.................................................Passed
Fix End of Files.........................................................Passed
Check Yaml...............................................................Passed
isort....................................................................Passed
black....................................................................Passed
blackdoc.................................................................Passed
flake8...................................................................Failed
- hook id: flake8
- exit code: 1

xgcm/grid_ufunc.py:399:5: F821 undefined name 'reveal_type'
xgcm/grid_ufunc.py:554:5: F821 undefined name 'reveal_type'

mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

xgcm/grid_ufunc.py:399: note: Revealed type is "builtins.tuple[Any]"
xgcm/grid_ufunc.py:554: note: Revealed type is "builtins.list[Any]"
```

Reply:
@dcherian do you have any idea what I would need to change to get `mypy` to understand that 

```python
@as_grid_ufunc()
def ufunc1(a: Annotated[str, "(X:center)"]):
    ...
```
should raise a typing error? (Because `str` is not `np.ndarray`)

Reply:
Sorry no idea.